### PR TITLE
[ENH] Export and import graph classes from other common interfaces

### PR DESCRIPTION
**Pull request by adam2392**
Towards #57 

Changes proposed in this pull request:
- numpy export
- pcalg export
- tetrad export
- causal learn export now tested
- I removed "integration tests", since pywhy-graphs is more of an upstream package and we want to integrate support for exporting/importing other graph packages

Now need to add manual tests using what is present in: https://github.com/py-why/causal-learn/blob/main/tests/fci-test-data/result-py-FCI-8.txt for tetrad.

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by jaron-lee** [comment]
At some point it would be great if you could pick one of `causallearn.py, tests/test_causallearn.py` or `causal_learn.py, tests/test_causal_learn.py`, it's ruining my Vim projectionist heuristics when I switch back and forth 😂

**Comment by jaron-lee** [comment]
So I notice that `tigramite` has a hard dependency on exactly one patch version of scipy (`1.8.0`) - so happens that ananke requires `^1.8.1`. In this particular case I could go and decrement the patch version in ananke, but is there another way to set this up in poetry? The version conflict here is not that bad given that we don't actually need both of these packages at the same time per se.

**Comment by jaron-lee** [comment]
Ran into another issue - when I try and run unit tests python is picking up `networkx.py` as the `networkx` package (instead of whatever comes from `pip`). Is there a way to distinguish between the local module vs the pip package?

**Comment by adam2392** [comment]
What's the issue? The export submodule or the networkx submodule for mixed edge graphs?

**Comment by adam2392** [comment]
> So I notice that `tigramite` has a hard dependency on exactly one patch version of scipy (`1.8.0`) - so happens that ananke requires `^1.8.1`. In this particular case I could go and decrement the patch version in ananke, but is there another way to set this up in poetry? The version conflict here is not that bad given that we don't actually need both of these packages at the same time per se.

Ah yeah that's an issue. If it's easy to do it for ananke, then that would be great! 

I'll also submit an issue to tigramite to make the dependency not hard. I know the maintainer.

**Comment by jaron-lee** [comment]
> What's the issue? The export submodule or the networkx submodule for mixed edge graphs?

I think the export submodule - basically running tests from that folder would get python confused about which networkx to import. Might just be a weird issue on my test runner's end tho.

**Comment by adam2392** [comment]
Ah I see. We can def rename things if it's an issue. 

You can just run pytest from the root of the package folder. E.g.

Pytest ./pywhy_graphs/...

And it should work.

**Comment by adam2392** [comment]
This is somewhat of a big PR, and happy to break it up, but just wanted to get a high level review of the code from you @jaron-lee and @robertness to make sure the API looks good to you?

File: pywhy_graphs/export/tetrad.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+    Notes
+    -----
+    See tetrad documentation: https://cmu-phil.github.io/tetrad/manual/.
+
+    Warning: There is no guarantee or check that your file adheres to the
+    tetrad format.
+    """
+    # instantiate the type of causal graph
+    if graph_type == "dag":
```
Comment (jaron-lee) [inline pywhy_graphs/export/tetrad.py:31]: I think I have seen different ways of dealing with DAGs - should we use nx.DiGraph or send it to pywhy_graphs? ADMG here seems like a hack. Maybe an in-house pywhy_graphs DAG class is needed?

File: pywhy_graphs/testing.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import networkx as nx
+
+
+def assert_mixed_edge_graphs_isomorphic(G1, G2):
```
Comment (jaron-lee) [inline pywhy_graphs/testing.py:4]: This looks good, I presume all roundtrip tests should use this testing function ?

File: pywhy_graphs/export/tests/test_tetrad.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import pywhy_graphs
+from pywhy_graphs.export import graph_to_tetrad, tetrad_to_graph
+from pywhy_graphs.testing import assert_mixed_edge_graphs_isomorphic
+
+root = Path(pywhy_graphs.__file__).parents[1]
+filename = root / "pywhy_graphs" / "export" / "tests" / "test_pag_tetrad.txt"
+
+
+def test_to_tetrad(tmp_path):
+    G = tetrad_to_graph(filename, graph_type="pag")
```
Comment (jaron-lee) [inline pywhy_graphs/export/tests/test_tetrad.py:12]: What happens if a user has `graph_type="dag"` instead here? Will this function detect that there are `o->` edges and prevent that? Clearly the imported file does not represent a DAG so we shouldn't let the user make this mistake. Or we should have clear tests to show what happens when you import a graph under a different graph assumption than the one it was created with

**Comment by jaron-lee** [comment]
> > In the `<package>_to_graph` function we have this argument `graph_type`. Should this be a mandatory argument? Sometimes (e.g. in Ananke) the graph already carries a type which has a corresponding equivalent in pywhy_graphs, so `graph_type` can be easily inferred.
> 
> The `graph_type` parameter is just used to make sure the user explicitly spells out which graph to convert to because sometimes it is not clear. For example, pcalg encodes CPDAG and PAG as a numpy array, but there 1s and 0s mean different things, so we need to have different logic. If ananke can let us know easily which graph type to use, then I would say the `graph_type` parameter is not needed.
> 
> Note: not all graph types are supported in each export function. E.g. pcalg. We can change that tho, but I avoided the extra work.
> 
> > Additionally the `graph_to_<package>` function is typed for `MixedEdgeGraph` input. Do we expect users to actually pass in `MixedEdgeGraph`, or will we expect them to pass a child class like `ADMG`? Or should we expect both? This also has implications for exporting to Ananke.
> 
> `graph_to_<package>` can usually eat any graph because we have semantic separation of all types of graphs. The ADMG subclasses a MixedEdgeGraph but with very specific structure.
> 
> Lmk how that sounds. If there's anything you want me to change for downstream ananke PR, lmk!

This all looks good. What I'll do is if a user passes in something of type `MixedEdgeGraph` then I'll just handle it equivalently on the Ananke side (`Graph`) without doing any class inferring.

**Review by jaron-lee — APPROVED** [review]

File: pywhy_graphs/export/tetrad.py (hunk 1 above)
Comment (adam2392) [reply pywhy_graphs/export/tetrad.py:31]: True. Perhaps `nx.DiGraph` with some added semantics to ensure construction is acyclic... But it'd be too expensive to check acyclicity every time someone adds a node. 

On one hand it seems overkill to recreate a DAG class when nx.DiGraph works. On the other hand, the mental switching is a bit weird. 

Tbh not sure exactly which way to go.

File: pywhy_graphs/testing.py (hunk 2 above)
Comment (adam2392) [reply pywhy_graphs/testing.py:4]: Yep!

FYI I realized there's a lot of LOC scattered in the codebase doing this.

File: pywhy_graphs/export/tests/test_tetrad.py (hunk 3 above)
Comment (adam2392) [reply pywhy_graphs/export/tests/test_tetrad.py:12]: I'm going to leave this for another PR.

**Pull request merged**